### PR TITLE
Updated version skew strategy for InPlacePodVerticalScaling

### LINKED ISSUE #117767: [FG:InPlacePodVerticalScaling] Implement version skew handling for in-place pod resize
### What would you like to be added?

This is a tracking bug to implement two changes to better handle version skew as documented in https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#version-skew-strategy
1. kubelet: When feature-gate is disabled, if kubelet sees a Proposed resize, it rejects the resize as Infeasible.
2. scheduler: If PodStatus.Resize field is not empty, scheduler uses max(ResourcesAllocated, Requests) even if the feature-gate is disabled.

### Why is this needed?

This is needed to gracefully handle component version skews that can occur during upgrades.

### PR DESCRIPTION
**Pull request by sreeram-venkitesh**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Updates the version skew strategy for InPlacePodVerticalScaling for beta graduation.

#### Which issue(s) this PR fixes:

Fixes #117767

#### Does this PR introduce a user-facing change?

```release-note
Version skew strategy update for InPlacePodVerticalScaling for beta graduation.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/1287
```

### CODE REVIEW

**Comment by sreeram-venkitesh** [comment]
@tallclair Please let me know if I’m in the right direction. Some tests are breaking, I’m still working on fixing them. This is how I tested the change:

- Create kind cluster with 3 nodes: 
  - control plane with `kindest/node:latest` image and InPlacePodVerticalScaling enabled
  - worker # 1 with `kindest/node:v1.31.0` with InPlacePodVerticalScaling disabled
  - worker # 2 with `kindest/node:v1.30.0` with  InPlacePodVerticalScaling disabled

Previous to these changes, I was able to update Pod's resources. In v1.30, the Pod gets restarted, and in v1.31, the Pod gets updated, but its not restarted so the resource update is not reflected until the next restart. After the changes I made in this PR, Pod resource update is not allowed for Pods created on both the worker nodes with InPlacePodVerticalScaling disabled. The validation checks if the Pod status has resources defined before update and I’ve used the same logic you wrote in the KEP.  

This is how the error looks like right now when you try to update a Pod resource without enabling the feature gate. I’m also getting the error for how Resources cannot be updated along with this. I’m assuming this is because `cmp.Diff(oldPod.Spec, mungedPodSpec)` is non-empty. I’m looking into this as well. Also the string `Pod running on node without IPPVS enabled may not be updated` is a placeholder, I will change this.

```yaml
# pods "nginx-pod" was not valid:
# * spec: Forbidden: Pod running on node without IPPVS enabled may not be updated
# * spec: Forbidden: pod updates may not change fields other than `spec.containers[*].image`,`spec.initContainers[*].image`,`spec.activeDeadlineSeconds`,`spec.tolerations` (only additions to existing tolerations),`spec.terminationGracePeriodSeconds` (allow it to be set to 1 if it was previously negative)
#   core.PodSpec{
#       Volumes:        {{Name: "kube-api-access-njhml", VolumeSource: {Projected: &{Sources: {{ServiceAccountToken: &{ExpirationSeconds: 3607, Path: "token"}}, {ConfigMap: &{LocalObjectReference: {Name: "kube-root-ca.crt"}, Items: {{Key: "ca.crt", Path: "ca.crt"}}}}, {DownwardAPI: &{Items: {{Path: "namespace", FieldRef: &{APIVersion: "v1", FieldPath: "metadata.namespace"}}}}}}, DefaultMode: &420}}}},
#       InitContainers: nil,
#       Containers: []core.Container{
#               {
#                       ... // 6 identical fields
#                       EnvFrom: nil,
#                       Env:     nil,
#                       Resources: core.ResourceRequirements{
#                               Limits: nil,
#                               Requests: core.ResourceList{
#                                       s"cpu":    {i: {...}, s: "10m", Format: "DecimalSI"},
# -                                     s"memory": {i: resource.int64Amount{value: 104857600}, s: "100Mi", Format: "BinarySI"},
# +                                     s"memory": {i: resource.int64Amount{value: 105906176}, s: "101Mi", Format: "BinarySI"},
#                               },
#                               Claims: nil,
#                       },
#                       ResizePolicy:  {{ResourceName: s"cpu", RestartPolicy: "NotRequired"}, {ResourceName: s"memory", RestartPolicy: "NotRequired"}},
#                       RestartPolicy: nil,
#                       ... // 13 identical fields
#               },
#       },
#       EphemeralContainers: nil,
#       RestartPolicy:       "Always",
#       ... // 28 identical fields
#   }
```

File: pkg/apis/core/validation/validation.go
Comment (tallclair) [inline pkg/apis/core/validation/validation.go:5176]: This will probably be OK once we switch to the resize subresource (https://github.com/kubernetes/kubernetes/issues/109553), but to call it here you'll need to confirm it's actually a resize request (there are other valid pod updates, including metadata & image)

File: pkg/apis/core/validation/validation.go
Comment (tallclair) [inline pkg/apis/core/validation/validation.go:5186]: I think skip this check here. The error will already be added above, so flagging resource mutation as invalid here will just make for a confusing and redundant error message.

File: pkg/apis/core/validation/validation.go
Comment (tallclair) [inline pkg/apis/core/validation/validation.go:5173]: this might need to move to `PodValidationOptions`. TBH, I'm not sure what decides what goes in the validation options - maybe ask an API reviewer.

File: pkg/apis/core/validation/validation.go
Comment (tallclair) [inline pkg/apis/core/validation/validation.go:5302]: Add a comment here explaining what this is doing and why (i.e. for version skew, and why it works)

Comment (tallclair) [reply pkg/apis/core/validation/validation.go:5302]: Also, leave a TODO to remove this code after GA+3 releases.

**Review by tallclair — COMMENTED** [review]
On the right track. Thanks!

File: pkg/apis/core/validation/validation.go
Comment (sreeram-venkitesh) [reply pkg/apis/core/validation/validation.go:5173]: I was trying to move this `isPodResizeRequestValid` function to PodValidationOptions and make it aligned with everything else there, but I soon realized that the opts that `ValidatePodUpdate` gets comes from the `GetValidationOptionsFromPodSpecAndMeta` function. We need to check if Resources are present in the Status however. Can PodValidationOptions have logic based on the PodStatus?

https://github.com/kubernetes/kubernetes/blob/8c7160205db67b250dece7cad5c6367debad2642/pkg/registry/core/pod/strategy.go#L149-L151

File: pkg/apis/core/validation/validation.go
Comment (sreeram-venkitesh) [reply pkg/apis/core/validation/validation.go:5176]: @tallclair In order to test this, I created a new branch from Anish's PR and added these changes to the new `ValidatePodResize` function that is added. This way this validation error would be added only to resize requests and all the other tests wouldn't fail like in this PR. Once Anish's PR gets merged I can move these changes to the `ValidatePodResize` function too. I opened https://github.com/kubernetes/kubernetes/pull/128540 to test this first and be ready before code freeze. I'll close it once I'm done testing.

Comment (tallclair) [reply pkg/apis/core/validation/validation.go:5176]: Yes, agreed - once Anish's PR merges this should just move to `ValidatePodResize`, and you don't need to worry about the feature gate.

**Review by tallclair — COMMENTED** [review]
I think this is good to go once you rebase it on the /resize subresource PR.

Have you manually tested the changes? If not, I can share some instructions for setting up a version-skewed Kind cluster to test locally.

**Comment by sreeram-venkitesh** [comment]
@tallclair I've rebased my changes and squashed everything into a single commit, PTAL!

> Have you manually tested the changes?

I had tested the skew before the rebase as described in https://github.com/kubernetes/kubernetes/pull/128186#issuecomment-2423180494. I will test it once again today.

File: pkg/apis/core/validation/validation.go
Comment (tallclair) [inline pkg/apis/core/validation/validation.go:5564]: ValidatePodResize is never called if the feature gate isn't enabled, so no need to check it here.

File: pkg/apis/core/validation/validation.go
Comment (tallclair) [inline pkg/apis/core/validation/validation.go:5555]:
```suggestion
// isPodResizeRequestSupported checks whether the pod being is running on a node with InPlacePodVerticalScaling enabled. 
func isPodResizeRequestSupported(pod core.Pod) bool {
```

File: pkg/apis/core/validation/validation.go
Comment (tallclair) [inline pkg/apis/core/validation/validation.go:5570]:
```suggestion
	// No running containers. We cannot tell whether the node supports resize at this point, so we assume it does.
```

**Review by tallclair — COMMENTED** [review]
Thanks!

File: pkg/apis/core/validation/validation.go
Comment (tallclair) [inline pkg/apis/core/validation/validation.go:5511]: nit: no need for the variable
```suggestion
	if !isPodResizeRequestSupported(*oldPod) {
```

File: pkg/apis/core/validation/validation.go
Comment (tallclair) [inline pkg/apis/core/validation/validation.go:5512]:
```suggestion
		allErrs = append(allErrs, field.Forbidden(specPath, "Pod running on node without support for resize"))
```

File: pkg/apis/core/validation/validation_test.go
Comment (tallclair) [inline pkg/apis/core/validation/validation_test.go:25100]: nit: remove this

File: pkg/apis/core/validation/validation_test.go
Comment (tallclair) [inline pkg/apis/core/validation/validation_test.go:25216]: no longer necessary

**Review by tallclair — COMMENTED** [review]
Looks like the unit test needs to be fixed too.

**Comment by tallclair** [comment]
Wait, where'd the unit tests go? We don't need to change the feature gate, since this function isn't checking the feature gate, but we still want to manually test the logic. I'd expect to see test cases with & without running containers, and with & without status resources for each.

**Comment by sreeram-venkitesh** [comment]
I'll push them now.

**Comment by tallclair** [comment]
/lgtm

/assign @thockin 
For validation approval.

**Comment by tallclair** [comment]
Test failures look like unrelated flakes. That suite isn't pr blocking.

File: pkg/apis/core/validation/validation.go
Comment (thockin) [inline pkg/apis/core/validation/validation.go:5518]: @ndixita you have similar logic (different reason, same result) - PTAL and see if we can join them, whoever merges last :)

**Review by thockin — APPROVED** [review]
Thanks!

/lgtm
/approve

**Comment by pacoxu** [comment]
/milestone v1.32
approved before code freeze
/skip

**Pull request merged**
